### PR TITLE
Ensure tables can be created also when explicitly setting units to dimensionless

### LINKED ISSUE #17047: QTable cannot take `dimensionless_unscaled` when creating table from `data`
### Description

Attempting to create an table with a column in units `dimensionless_unscaled` leads to a crash, because a guard against setting a column attribute to `np.ma.masked` or `None` triggers a comparison between `np.ma.masked` and Unit which eventually resolves to a `ZeroDivisionError`.

### Expected behavior

A `QTable` is created. 

### How to Reproduce

```python
from astropy.table import QTable, Column
import numpy as np
data = np.ones((5,2))
tt = QTable(data=data, names=["a","weight"],units={"weight":u.dimensionless_unscaled})
```
Results in 
```
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
Cell In[15], line 2
      1 data = np.ones((5,2))
----> 2 tt = QTable(data=data, names=["a","weight"],units={"weight":u.dimensionless_unscaled})
      3 tt

File ~/miniconda3/envs/datapipe-testbench/lib/python3.11/site-packages/astropy/table/table.py:887, in Table.__init__(self, data, masked, names, dtype, meta, copy, rows, copy_indices, units, descriptions, **kwargs)
    884 if self.masked not in (None, True, False):
    885     raise TypeError("masked property must be None, True or False")
--> 887 self._set_column_attribute("unit", units)
    888 self._set_column_attribute("description", descriptions)

File ~/miniconda3/envs/datapipe-testbench/lib/python3.11/site-packages/astropy/table/table.py:921, in Table._set_column_attribute(self, attr, values)
    918     if value.strip() == "":
    919         value = None
--> 921 if value not in (np.ma.masked, None):
    922     col = self[name]
    923     if attr == "unit" and isinstance(col, Quantity):
    924         # Update the Quantity unit in-place

File ~/miniconda3/envs/datapipe-testbench/lib/python3.11/site-packages/numpy/ma/core.py:4182, in MaskedArray.__eq__(self, other)
   4171 def __eq__(self, other):
   4172     """Check whether other equals self elementwise.
   4173 
   4174     When either of the elements is masked, the result is masked as well,
   (...)
   4180     and other considered equal only if both were fully masked.
   4181     """
-> 4182     return self._comparison(other, operator.eq)

File ~/miniconda3/envs/datapipe-testbench/lib/python3.11/site-packages/numpy/ma/core.py:4138, in MaskedArray._comparison(self, other, compare)
   4134 else:
   4135     # For regular arrays, just use the data as they come.
   4136     sdata = self.data
-> 4138 check = compare(sdata, odata)
   4140 if isinstance(check, (np.bool_, bool)):
   4141     return masked if mask else check

File ~/miniconda3/envs/datapipe-testbench/lib/python3.11/site-packages/astropy/units/core.py:955, in UnitBase.__eq__(self, other)
    952     return NotImplemented
    954 try:
--> 955     return is_effectively_unity(self._to(other))
    956 except UnitsError:
    957     return False

File ~/miniconda3/envs/datapipe-testbench/lib/python3.11/site-packages/astropy/units/core.py:1157, in UnitBase._to(self, other)
   1147     # Check quickly whether equivalent.  This is faster than
   1148     # `is_equivalent`, because it doesn't generate the entire
   1149     # physical type list of both units.  In other words it "fails
   1150     # fast".
   1151     if self_decomposed.powers == other_decomposed.powers and all(
   1152         self_base is other_base
   1153         for (self_base, other_base) in zip(
   1154             self_decomposed.bases, other_decomposed.bases
   1155         )
   1156     ):
-> 1157         return self_decomposed.scale / other_decomposed.scale
   1159 raise UnitConversionError(f"'{self!r}' is not a scaled version of '{other!r}'")

ZeroDivisionError: float division by zero
```

### Versions

```python
import astropy
try:
    astropy.system_info()
except AttributeError:
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import astropy; print("astropy", astropy.__version__)
    import numpy; print("Numpy", numpy.__version__)
    import erfa; print("pyerfa", erfa.__version__)
    try:
        import scipy
        print("Scipy", scipy.__version__)
    except ImportError:
        print("Scipy not installed")
    try:
        import matplotlib
        print("Matplotlib", matplotlib.__version__)
    except ImportError:
        print("Matplotlib not installed")
```
```
Linux-5.15.0-119-generic-x86_64-with-glibc2.35
Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:53:32) [GCC 12.3.0]
astropy 5.3.4
Numpy 1.26.4
pyerfa 2.0.0
Scipy 1.11.4
Matplotlib 3.8.3
```

### PR DESCRIPTION
**Pull request by mhvk**
This pull request is to address the problem encountered in #17047, where a `QTable` could not be created if a `u.dimensionless_unscaled` unit was passed in. While a simple fix to `Table` fixed the problem for tables (first pair of commits), it turned out there was also a somewhat subtle bug in units, which is that it was possible to create a unit with zero scale - which is a bad unit since nothing can be converted to it (scale ratios give `ZeroDivisionError`). So, the second pair of commits adds a test and a fix for that.

Fixes #17047

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/table/tests/test_table.py
Comment (eerovaher) [inline astropy/table/tests/test_table.py:3546]: This could fit on a single line:
```suggestion
    tt = QTable({"a": [1.0, 2.0, 3.0]}, units={"a": u.one})
```

File: astropy/table/table.py
Comment (nstarman) [inline astropy/table/table.py:941]: Is this stylistic or the source of a subtle bug?

Comment (nstarman) [reply astropy/table/table.py:941]: Oh, I see the test below. Tricky!

Comment (mhvk) [reply astropy/table/table.py:941]: It is the source of the bug: `not in list` calls `tuple.__contains__(value)` so the phrasing here is equivalent to `not any(value == t for t in (...))`, while what we want is `not any(value is t for t in (...))` (but for two elements, we might as well expand).

Note that the reason of the failure is somewhat subtle still: `np.ma.masked == unit` drops the mask to check equality of the underlying data. Since `np.ma.masked = MaskedArray(0., mask=True)`, this leads to `np.array(0) == unit`.  Arrays are greedy, so this becomes `np.array(0.) == np.array(unit, dtype=object)`. In the `dtype=object` loop of the `np.equal` ufunc, the individual items get compared, so one ends up with `0. == unit` which failed as `Unit(0.)` creates a unit with zero scale, and in trying to check whether that is the same as `u.one = Unit(1.)`, it is checked whether after decomposing to the same bases (which works), the ratio of scales is unity, so `1./0.` is calculated.

Indeed, the unit fix which forbids zero scale would also solve the table problem on its own, but since `None` and `np.ma.masked` are really meant to be used as singletons, the replacement is still appropriate (and maybe avoids other possibly problematic `value` instances...)

Comment (pllim) [reply astropy/table/table.py:941]: Is swapping to `(None, np.ma.masked)` not gonna work? Should we add a comment so future dev won't try to revert this when they attempt to "clean the code"?

File: astropy/table/tests/test_table.py
Comment (mhvk) [reply astropy/table/tests/test_table.py]: Thanks, better indeed. Two changes holding on is enough, so I pushed it up.

File: astropy/table/table.py
Comment (mhvk) [reply astropy/table/table.py:941]: > Is swapping to `(None, np.ma.masked)` not gonna work?

That would not matter. The comparison with `None` does reliably fail, but then one still has to compare with `np.ma.masked` - I swapped them in my implementation just because `None` seemed more likely to happen than `np.ma.masked` -- though of course almost always `value` is not going to be either of those.

File: astropy/units/tests/test_units.py
Comment (pllim) [inline astropy/units/tests/test_units.py:1044]: Do we need `allclose` or this is always going to be exact across all the archs?

File: astropy/table/table.py
Comment (mhvk) [reply astropy/table/table.py:941]: Whether a comment would be a good idea: not sure, in the end I felt that the moment it is written as `is`, the code reader will not change it to `not in` because then you do think that uses equality. The problem is really the other way around, that if you want to check whether something is in a list, `not in` seems so logical to use. But happy to add a comment if that is preferred!

Comment (pllim) [reply astropy/table/table.py:941]: No, it's okay. With the new tests, I think it is going to be obvious because the tests will fail again. Sorry for the noise... I didn't think it through when I wrote the comment above.

File: astropy/units/tests/test_units.py
Comment (mhvk) [reply astropy/units/tests/test_units.py:1044]: I picked a power of 2 on purpose...

**Review by pllim — APPROVED** [review]
I am not any of those subpackage maintainer but LGTM, so approving. I'll leave it to the author to decide if my approval is good enough or want more approvals. Thanks!

**Comment by mhvk** [comment]
@nstarman, @eerovaher - good to go in? I guess from a larger perspective, the only question is whether explicitly forbidding units with a scale of zero is the way to go (I can see no possible use, but perhaps lack imagination!).

File: astropy/units/core.py
Comment (eerovaher) [inline astropy/units/core.py:2322]: Wouldn't it be better if the code here called https://github.com/astropy/astropy/blob/32a6b7582f3029a65adc960f605d33663652211f/astropy/units/utils.py#L195
and it would be the job of that function to check if `scale` is zero? There's a `sanitize_scale()` call later on that we wouldn't need anymore.

Comment (mhvk) [reply astropy/units/core.py:2322]: I like the idea, but I don't think we can get rid of the later calls: both in the short-cut in `__init__` and inside `_expand_and_gather`, the scale is multiplied by other scales raised to some power, which makes it possible that one gets, e.g., a result very close to unity that should be passed through `_sanitize_scale`. That said, in those places, some of the tests on the scale are not needed (e.g., whether it has a `.dtype`). So, I think you are right that this can be refactored, with an early check on numpy scalars and value of zero, and then the checks for near-unity still in the same place. How about I do that in follow-up? I think it would go beyond what I would be comfortable backporting.

Comment (eerovaher) [reply astropy/units/core.py:2322]: What you've done is a minimal bugfix, and the practice of the project is indeed not to backport refactoring pull requests.

Comment (eerovaher) [reply astropy/units/core.py:2322]: Do you want me to have a look at the refactoring or would you prefer to do it yourself?

Comment (mhvk) [reply astropy/units/core.py:2322]: If you're interested, then please go ahead, thanks!

**Pull request merged**
